ls1intum/Artemis
# `Programming exercises`: Add possibility to import existing static code analysis configurations

# Note: This PR is based on #6026

### Checklist
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context
Closes #5169

Typically, Instructors want to use one standard SCA configuration for all exercises. This allows consistency in what type of feedback students get to see and how many points these issues deduct. Currently, setting up this configuration is cumbersome, since every category must be configured again and again for every exercise manually.

This PR introduces a new mechanism allowing editors to copy over an existing SCA configuration from another exercise. After defining this configuration once, you no longer need to define the same configuration again and again, but just import it into other exercises.

### Description

This PR introduces a new REST endpoint `/static-code-analysis-categories/import` which allows copying over the configuration from one exercise into another. The calling user must have at least editor rights for both exercises.

Additionally, this PR refactors the database calls when finding the available exercises on the different pages of the import dialog. Currently, for every possible filter combination and whether the current user is an admin or not, an own method in the different repositories exists (leading to 6 methods since the case of courseFilter == false and examFilter == false can be short-evaluated to just return an empty list). These methods were also duplicated in all 4 exercise repositories. With the addition of another `SCAFilter`. this approach was no longer maintainable. 

Instead, this PR introduces the [Criteria builder](https://docs.artemis.cit.tum.de/dev/guidelines/criteria-builder) which is already used for filtering posts. This allows us to dynamically create efficient queries reacting to the different filters. In order to reuse this implementation for all exercise types, a new `ExerciseSpecificationService` got added, which dynamically builds the specification based on the search string, the filters, and the settings (page number, page size, sorting).

### Steps for Testing
Prerequisites:
- 1 Instructor or Editor
- 2 Programming Exercises with SCA enabled
- 1 Programming Exercise with SCA deactivated

1. Setup one exercise to have a distinctive SCA configuration
2. Navigate to the grading page of the second exercise and press on the "Import Configuration" button
3. Check that only exercises with SCA active are available in the import dialog and that search / sort / and filters work.
4. Select one exercise and press import.
5. Verify that the other SCA categories got imported into the existing exercise.
6. Submit solutions as a Student and check that the configuration gets applied correctly.

#### Exam Mode Testing
Do the same steps as above. Verify that importing from and to exam exercises is possible.

### Review Progress

#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage
100% coverage for changed / new lines.
| Class/File | Branch | Line |
|------------|-------:|-----:|
| StaticCodeAnalysisService.java | 97% | 91% |
| StaticCodeAnalysisRessource.java | 100% | 90% |
| ExerciseSpecificationService.java | 100% | 100% |
| ProgrammingExerciseResource.java | 91% | 83% |

### Screenshots
TODO
